ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs

```diff
@@ -265,34 +272,113 @@ newResolverResource resolvConf = go
 --
 
 data TraceLocalRootPeers =
-       TraceLocalRootDomains [(DomainAddress, PeerAdvertise)]
+       TraceLocalRootDomains [(Int, Map RelayAddress PeerAdvertise)]
      | TraceLocalRootWaiting DomainAddress DiffTime
      | TraceLocalRootResult  DomainAddress [(IPv4, DNS.TTL)]
      | TraceLocalRootFailure DomainAddress DNSorIOError
        --TODO: classify DNS errors, config error vs transitory
   deriving Show
 
-
 -- |
 --
-localRootPeersProvider :: Tracer IO TraceLocalRootPeers
-                       -> TimeoutFn IO
-                       -> DNS.ResolvConf
-                       -> StrictTVar IO (Map DomainAddress (Map Socket.SockAddr PeerAdvertise))
-                       -> NonEmpty (DomainAddress, PeerAdvertise)
-                       -> IO Void
-localRootPeersProvider tracer timeout resolvConf rootPeersVar domains = do
-    traceWith tracer (TraceLocalRootDomains (NonEmpty.toList domains))
+localRootPeersProvider
+  :: Tracer IO TraceLocalRootPeers
+  -> TimeoutFn IO
+  -> DNS.ResolvConf
+  -> StrictTVar IO [(Int, Map Socket.SockAddr PeerAdvertise)]
+  -> StrictTVar IO [(Int, Map RelayAddress PeerAdvertise)]
+  -> IO Void
+localRootPeersProvider tracer
+                       timeout
+                       resolvConf
+                       rootPeersGroupsVar
+                       domainsGroupsVar = do
+  domainsGroups <- atomically $ readTVar domainsGroupsVar
+  traceWith tracer (TraceLocalRootDomains domainsGroups)
 #if !defined(mingw32_HOST_OS)
-    rr <- asyncResolverResource resolvConf
+  rr <- asyncResolverResource resolvConf
 #else
-    let rr = newResolverResource resolvConf
+  let rr = newResolverResource resolvConf
 #endif
-    withAsyncAll (map (monitorDomain rr) (NonEmpty.toList domains)) $ \asyncs ->
-      snd <$> waitAny asyncs
+  let
+      -- Flatten the local root peers groups and associate a group index to each
+      -- DomainAddress to be monitorized.
+      addresses = [ (group, address, pa)
+                  | (group, (_, m)) <- zip [0..] domainsGroups
+                  , (address, pa) <- Map.toList m ]
+      domains = [ (group, domain, pa)
+                | (group, RelayDomain domain, pa) <- addresses ]
+      ipsAddr = [ (group, IP.toSockAddr (ip, port), pa)
+                | (group, RelayAddress ip port, pa) <- addresses ]
+
+      -- Since we want to preserve the number of groups, the targets, and
+      -- the addresses within each group, we fill the TVar with
+      -- a placeholder list, in order for each monitored DomainAddress to
+      -- be updated in the correct group.
+      placeholder = map (\(t, _) -> (t, Map.empty)) domainsGroups
+      placeholder' =
+        foldr (\(group, addr, pa) ph
+                -> let (target, entry) = ph !! group
+                   in updateList group (target, Map.insert addr pa entry) ph)
+              placeholder
+              ipsAddr
```

Comment:
I wonder if we could simplify this with something like [`mapKeysWith`](https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Map-Internal.html#v:mapKeysWith).

Reply:
I'm not sure how

Reply:
Would this work: [this](https://github.com/input-output-hk/ouroboros-network/blob/coot/bolt12/master-local-root-providers/ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs#L330).

Reply:
We don't need to use an `IntMap` if we don't modify the spine of the list.

Reply:
That would not work because the `Int` in `[(Int, Map RelayAddress PeerAdvertise)]` is not unique and 2 different groups might have the same valency

Reply:
Ach right the `Int` is the target; what about:
```
rootPeersGroups = map (\(target, m) -> (target, f m)) domainsGroups
  -- where `f` the same as in my branch
```

Reply:
Ah yes I think that would work! :D